plugins/amqp_compressed_proof_plugin/amqp_compressed_proof_plugin.cpp

```diff
@@ -0,0 +1,393 @@
+#include <eosio/chain/exceptions.hpp>
+
+#include <eosio/reliable_amqp_publisher/reliable_amqp_publisher.hpp>
+
+#include <boost/multi_index_container.hpp>
+#include <boost/multi_index/key.hpp>
+#include <boost/multi_index/ordered_index.hpp>
+#include <boost/multi_index/hashed_index.hpp>
+#include <boost/algorithm/string/predicate.hpp>
+#include <boost/algorithm/string.hpp>
+
+#include <fc/io/cfile.hpp>
+
+//using boost::multi_index_container;
+using namespace boost::multi_index;
+using namespace std::string_literals;
+
+namespace eosio {
+static appbase::abstract_plugin& _amqp_compressed_proof_plugin = app().register_plugin<amqp_compressed_proof_plugin>();
+
+struct action_entry {
+   chain::action         action;
+   std::vector<char>     return_value;
+   chain::action_receipt action_receipt;
+
+   action_entry() = default;
+
+   action_entry(const chain::action_trace& atrace) :
+     action(atrace.act), return_value(atrace.return_value), action_receipt(*atrace.receipt) {}
+
+   bool operator<(const action_entry& other) const {
+      return action_receipt.global_sequence < other.action_receipt.global_sequence;
+   }
+};
+
+struct combine_nodes{};
+using merkle_cmd = fc::static_variant<chain::digest_type, action_entry, combine_nodes>;
+
+struct merkle_creator_node {
+   chain::checksum_type  hash;
+   std::list<merkle_cmd> cmd_stream;
+};
+
+class merkle_cmd_creator {
+public:
+   void add_noninterested_action(const action_entry& ae) {
+      handle_input_action(merkle_creator_node{chain::digest_type::hash(ae.action_receipt)});
+   }
+
+   void add_interested_action(const action_entry& ae) {
+      handle_input_action(merkle_creator_node{chain::digest_type::hash(ae.action_receipt), {ae}});
+   }
+
+   std::vector<char> generate_serialized_proof(bool arv_activated) {
+      if(!first_input_pair && curent_level.empty())
+         return std::vector<char>();
+      if(first_input_pair && curent_level.empty())
+         return fc::raw::pack<bool, std::list<merkle_cmd>>((bool)arv_activated, first_input_pair->cmd_stream);
+
+
+      if(first_input_pair) {
+         merkle_creator_node even_pair{first_input_pair->hash};
+         curent_level.emplace_back(combine(*first_input_pair, even_pair));
+      }
+
+      while(curent_level.size() > 1) {
+         if(curent_level.size() % 2)
+            curent_level.emplace_back(merkle_creator_node{curent_level.back().hash});
+         for(unsigned int i = 0; i < curent_level.size(); i+=2)
+            curent_level[i/2] = combine(curent_level[i], curent_level[i+1]);
+         curent_level.resize(curent_level.size()/2);
+      }
+
+      return fc::raw::pack<bool, std::list<merkle_cmd>>((bool)arv_activated, curent_level[0].cmd_stream);
+   }
+
+private:
+   void handle_input_action(merkle_creator_node input) {
+      if(!first_input_pair) {
+         first_input_pair = input;
+         return;
+      }
+
+      curent_level.emplace_back(combine(*first_input_pair, input));
```

Comment:
Is there a reason for combining the leaf nodes in `handle_input_actions`?  It would simplify the code a bit if `generate_serialized_proof` handled all the combining.

Reply:
I think there was a reason in a previous impl. But indeed now none of that is needed, can just be chopped out.